### ISSUE TITLE
ValueBinder error in a complex query

### ISSUE THREAD

**Issue by greew** [cakephp/cakephp#8796]
This is a (multiple allowed):
* [x] bug
* [ ] enhancement
* [ ] feature-discussion (RFC)

* CakePHP Version: 3.2.8
* Platform and Target: OS X, PHP 7.0.6, MySQL 5.7.11

### What you did
A huge query of mine resulted in an invalid MySQL query, and I've now been able to reproduce it in a smaller environment, based on the cakephp/app structure, see greew/cakephp-orm-error@4a82c8ec29577171c14dd4c256c06294f2fdf147

    ATable hasMany BTable
    BTable hasMany CTable (using subquery strategy!!)

`BTable` has a `beforeFind` method (in my huge query, it's a behaviour, but to minify the test app, I've added it directly in the model), which adds a subquery to the `select` part of the query. 
This subquery works fine, but it adds some data to the ValueBinder of the query before the ValueBinder-stuff from the master query is added. 

When adding

    var_dump($statement, !$isObject ?: $query->valueBinder()->bindings());

before https://github.com/cakephp/cakephp/blob/6a463bbec5f3cf71de2a9f1a42b1ab71ba0f8fa0/src/Database/Driver/Mysql.php#L122

I get the following output:

```
/Users/jesper/code/test/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:122:

object(PDOStatement)[120]
  public 'queryString' => string 'SELECT A.id AS `A__id`, A.name AS `A__name` FROM a A WHERE name = :c0' (length=69)

/Users/jesper/code/test/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:122:
array (size=1)
  ':c0' => 
    array (size=3)
      'value' => string 'first' (length=5)
      'type' => string 'string' (length=6)
      'placeholder' => string 'c0' (length=2)

/Users/jesper/code/test/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:122:
object(PDOStatement)[166]
  public 'queryString' => string 'SELECT (SELECT (CONCAT((user_id), :c0, (created))) FROM audit_logs AuditLogs WHERE (foreign_key = (B.id) AND model = :c1) LIMIT 1) AS `created` FROM b B WHERE B.a_id in (:c2)' (length=174)

/Users/jesper/code/test/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:122:
array (size=3)
  ':c0' => 
    array (size=3)
      'value' => string ',' (length=1)
      'type' => null
      'placeholder' => string 'c0' (length=2)
  ':c1' => 
    array (size=3)
      'value' => string 'B' (length=1)
      'type' => string 'string' (length=6)
      'placeholder' => string 'c1' (length=2)
  ':c2' => 
    array (size=3)
      'value' => int 1
      'type' => string 'integer' (length=7)
      'placeholder' => string 'c2' (length=2)

/Users/jesper/code/test/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:122:
object(PDOStatement)[180]
  public 'queryString' => string 'SELECT C.id AS `C__id`, C.b_id AS `C__b_id`, C.name AS `C__name` FROM c C INNER JOIN (SELECT (B.id) FROM b B WHERE B.a_id in (:c0) GROUP BY B.id ) B ON C.b_id = (B.id)' (length=167)

/Users/jesper/code/test/vendor/cakephp/cakephp/src/Database/Driver/Mysql.php:122:
array (size=1)
  ':c0' => 
    array (size=3)
      'value' => string ',' (length=1)
      'type' => null
      'placeholder' => string 'c0' (length=2)
```

The above shows, that the last query, which is loaded with subquery strategy, is using the first (:c0) ValueBinder placeholder instead of the third (:c2) which contains the correct data.

It ends out in the following queries, the first two being correct, the last one asking for `B.a_id in (',')` instead of `B.a_id in (1)`:

```sql
SELECT A.id AS `A__id`, A.name AS `A__name` FROM a A WHERE name = 'first'

SELECT (SELECT (CONCAT((user_id), ',', (created))) FROM audit_logs AuditLogs WHERE (foreign_key = (B.id) AND model = 'B') LIMIT 1) AS `created` FROM b B WHERE B.a_id in (1)

SELECT C.id AS `C__id`, C.b_id AS `C__b_id`, C.name AS `C__name` FROM c C INNER JOIN (SELECT (B.id) FROM b B WHERE B.a_id in (',') GROUP BY B.id ) B ON C.b_id = (B.id)
```

**Comment by greew** [comment]
Whoops... forgot structure of the tables:

```sql
-- Create syntax for TABLE 'a'
CREATE TABLE `a` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `name` varchar(255) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8;

-- Create syntax for TABLE 'audit_logs'
CREATE TABLE `audit_logs` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `user_id` int(11) unsigned NOT NULL,
  `created` datetime NOT NULL,
  `model` char(1) NOT NULL DEFAULT '',
  `foreign_key` int(11) unsigned NOT NULL,
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=2 DEFAULT CHARSET=utf8;

-- Create syntax for TABLE 'b'
CREATE TABLE `b` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `a_id` int(11) unsigned NOT NULL,
  `name` varchar(255) NOT NULL DEFAULT '',
  `type` char(1) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=13 DEFAULT CHARSET=utf8;

-- Create syntax for TABLE 'c'
CREATE TABLE `c` (
  `id` int(11) unsigned NOT NULL AUTO_INCREMENT,
  `b_id` int(11) unsigned NOT NULL,
  `name` varchar(255) NOT NULL DEFAULT '',
  PRIMARY KEY (`id`)
) ENGINE=InnoDB AUTO_INCREMENT=4 DEFAULT CHARSET=utf8;
```

**Comment by greew** [comment]
As far as I can see, the problem arises in https://github.com/cakephp/cakephp/blob/2f54ee3ca0aeba3e235119cd61df1808b6d54a03/src/ORM/Association/SelectableAssociationTrait.php#L233, where we modify the select fields without checking that there are no placeholders within the previous select fields. 

If we checked this first and modified the ValueBinder accordingly, it might solve the problem.

I'll give it a try and see, if I can come up with something.

**Comment by greew** [comment]
Closed as PR #8805 is now open.
